### PR TITLE
Attempt at a fix for issue #176

### PR DESCRIPTION
**Pull request by unsynchronized**
When "Darken when checked" is enabled, footprints that match the
criterion chosen (i.e. sourced or placed) are highlighted in a separate
color.  The currently-highlighted footprint is also shown in a third
color if it has met the sourced/placed criterion.

Upstream issue #176.

### CODE REVIEW

File: InteractiveHtmlBom/web/render.js [hunk 1]
```diff
@@ -342,16 +342,34 @@ function drawFootprints(canvas, layer, scalefactor, highlight) {
...
+          drawFootprint(ctx, layer, scalefactor, mod, bothPadColor, bothOutlinecolor, highlight, outline);
+        } else {
+          drawFootprint(ctx, layer, scalefactor, mod, highlightedPadColor, highlightedOutlinecolor, highlight, outline);
+        }
+      } else {
+        if(darkenedFootprints.has(i)) {
+          console.log("darkened " + darkenedPadColor + "  plain " + padcolor);
+          drawFootprint(ctx, layer, scalefactor, mod, darkenedPadColor, darkenedOutlinecolor, highlight, outline);
+        } else {
+          drawFootprint(ctx, layer, scalefactor, mod, padcolor, outlinecolor, highlight, outline);
```
Comment (qu1ck) [inline InteractiveHtmlBom/web/render.js:369]: This now draws a bounding box for every single footprint which is a bug.

In general the whole block has too many nested conditions.

I suggest following: 
* make padcolor and outlinecolor a single object passed into drawFootprint()
* precalculate 4 instances of that object based on style: normal, highlighted, darkened, both.
* write non-nested if () {} else if () {} ... statements that handle each case:
  * non-highlight mode: draw normal
  * highlightedFootprints and darkenedFootprints have the ref: draw with "both" color
  * darkenedFootprints has the ref: draw "darkened"
  * highlightedFootprints has the ref: draw "highlighted"

**Review by qu1ck — CHANGES_REQUESTED** [review]
Good for a first pass, needs some tweaks.

**Comment by qu1ck** [comment]
Let me know if you don't plan to address the review comments in, say, next week or two. I will touch this up and merge it.

**Comment by unsynchronized** [comment]
I will soon -- got busy, sorry!

**Comment by qu1ck** [comment]
Yeah, no rush, just wanted to know if it's still on your radar.

**Comment by nopeppermint** [comment]
@unsynchronized any update on this?

**Pull request closed without merging**